### DISCUSSION TITLE
Usage of snyk for vulnerability analysis.

### DISCUSSION THREAD

**Discussion by haydenmcp** [ghiscoding/lerna-lite#196] [General]
### Clear and concise description of the problem

Just a recommendation. No problem.

### Suggested solution

Visit [Snyk](https://snyk.io/) and sign your project up for vulnerability analysis if you don't already. It's free for public projects. Thanks for the awesome work you're doing!

### Validations

- [X] Follow our [Code of Conduct](https://github.com/ghiscoding/lerna-lite/blob/master/.github/CODE_OF_CONDUCT.md)
- [X] Read the [docs](https://github.com/ghiscoding/lerna-lite#readme).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by ghiscoding** [comment]
Converting to a Discussion since this seems to be more appropriate, thanks for the suggestion, I'll look into that later. I do know Snyk, though I didn't know we could register a lib. Thanks

**Comment by haydenmcp** [comment]
The reason I bring this up is because lerna v5 had some vulnerabilities present and I noticed a couple coming from this library, though it's a great library. It's just one more tool usable in increasing the value and reach of a project. :-)

**Reply by ghiscoding** [reply]
Isn't that doing something similar to what dependabot is currently doing? I assume that might install a bot that will raise PR if it finds any vulnerabilities? 

As per your comment, I did merge just a few minutes ago, some PRs that were opened in original Lerna v5 and replicated here to address some issues that they found while doing `npm audit` I think. 

Snyk seems to have problems trying to read all the `__fixtures__`, which he shouldn't read though, any ways to disable scanning those?  
![image](https://user-images.githubusercontent.com/643976/171020521-c637e030-16e5-474a-831c-275837e00700.png)

**Reply by haydenmcp** [reply]
You know, I assume they are doing the same thing. However, I've received alerts from Dependabot that I didn't from Snyk and vice versa. It's strange. And that's awesome to hear that you've resolved some of those issues.

**Reply by haydenmcp** [reply]
Check out the following: https://docs.snyk.io/products/snyk-code/getting-started-with-snyk-code/activating-snyk-code-using-the-web-ui/step-3-importing-repositories-to-snyk-for-the-snyk-code-testing/excluding-directories-and-files-from-the-import-process
That describes how to incorporate exclusions from the import process.

**Reply by ghiscoding** [reply]
I'd be happy to receive a PR for this if you want to contribute 😉 
I did add the project to Snyk, I think that part is done

**Reply by haydenmcp** [reply]
I'll take a look at it!

**Reply by haydenmcp** [reply]
@ghiscoding are you comfortable excluding all fixtures from the analysis?

**Reply by ghiscoding** [reply]
@haydenmcp yes I would suggest to exclude all `__fixtures__` since a few of them have `package.json` files which conflicts with Lerna-Lite itself and shouldn't be tested by Snyk since they are in place only for unit testing the lib. Thanks

**Reply by haydenmcp** [reply]
@ghiscoding Okay, cool. I actually don't need to do anything. When you add project in snyk you'll see a textbox at the bottom that says "exclude" or something to that effect. If you enter \_\_fixtures\_\_ in that box and then add the project it should import without issues based on my test.

**Reply by ghiscoding** [reply]
oh ok I see, I didn't scroll down neither read it all at first, too much info lol

![image](https://user-images.githubusercontent.com/643976/171056058-f676a780-18ab-4fd0-bb12-18a70904c7b0.png)

Now the only warnings I'm still getting, I just don't understand why it talks about a missing yarn lock file when the project is built with npm workspaces. I could probably just ignore them 
![image](https://user-images.githubusercontent.com/643976/171056557-d67043a8-edf0-427c-9b2e-57ab716c1a42.png)

**Reply by haydenmcp** [reply]
Yeah, the system operates using either a package.json, yarn.lock or it'll check the node_modules folder for dependencies. But that's really strange because I'm not seeing that error in my repos....do you use yarn?

**Reply by haydenmcp** [reply]
I'm wondering if the fact that I use yarn as my package manager and have a root-level yarn.lock is why I don't see that type of error.

**Reply by ghiscoding** [reply]
I use Yarn in other projects not this one, this one specifically uses NPM Workspace and has a [package-lock.json](https://github.com/ghiscoding/lerna-lite/blob/main/package-lock.json) in the root, so I wonder why it begs about a yarn lock file, I also did a search solution and the only places Yarn is reference is when Lerna deals with Yarn code explicitly and it's not in any `package.json` or anything, so it's kinda weird. Also if I made the mistake of using yarn instead of npm (I do sometime, old habits), then that would create a yarn lock file and I would know right away because that would become a changed file that would require commit, so I don't understand why it complains.... oh well

**Reply by haydenmcp** [reply]
Oh I didn't realize npm had workspaces and old habits die hard. :-) Do you happen to see those projects in any issues? Just want to make sure they're actually being picked up.

**Reply by ghiscoding** [reply]
Yeah npm as [workspaces](https://docs.npmjs.com/cli/v7/using-npm/workspaces) since npm@7 and I'm using in Lerna-Lite so that I don't need to add extra external dependencies and the `workspaces` in the package tells npm to build that in that exact order (or else the build would fail because for example the CLI needs to be run last because it depends on other packages). 

https://github.com/ghiscoding/lerna-lite/blob/66b756ad2e3765be16141e15e052d870d3b26646/package.json#L94-L105

> Do you happen to see those projects in any issues? Just want to make sure they're actually being picked up.

Sorry, I'm not sure what you mean in this question. The print screen I've shown you earlier is when I run "Add Projects" in Snyk, if that was your question

**Reply by haydenmcp** [reply]
I should have said packages. Hahaha. Sorry about that. If you navigate to lerna-lite on the snyk projects pane and you click on the project. A dropdown should show up with various vulnerabilities shown. If you click on a given path (mine shows paths to package.json files in my packages folder of my project), you'll see issues related to that package.json. Do you see any of those packages for which an error was shown in the log above in said drop-down?

**Reply by ghiscoding** [reply]
It actually doesn't show in the Project list, it seems that I need to add it every time, my other project shows in this list but not Lerna-Lite, however if I click on the "View last import log", I do see Lerna-Lite yarn lock file errors. So every time I want to test it, it seems that I need to click on Add Project over and over and also add `__fixtures__` every time. I'm not sure why it doesn't show in the list, it does say I have 2 project installed but shows only 1, I'm not sure if it's because of those error that it doesn't show in the list 

![image](https://user-images.githubusercontent.com/643976/171061936-6b7f8a45-3531-49a9-9454-ade2713557ec.png)

**Reply by haydenmcp** [reply]
Wow, that's crazy...if you reload the page it doesn't show up?

**Reply by haydenmcp** [reply]
What's strange about that is I forked lerna-lite and did the import without any problems except for a couple of errors related to malformed JSON! lol

**Reply by haydenmcp** [reply]
Do you see a list of package.json files?

**Reply by ghiscoding** [reply]
dohhh then I don't know, it seems we get different result lol

**Reply by haydenmcp** [reply]
Oh my gosh....I was wrong. I did get the same errors but I have \_\_fixtures\_\_ so that may be why I get plenty of package.jsons. Hmmmm....let me take another day to review this. This may be a problem with Snyk's current operation. For workspaces it may require a yarn.lock.

**Reply by haydenmcp** [reply]
Welp...npm workspaces aren't supported. :-/ https://docs.snyk.io/products/snyk-open-source/language-and-package-manager-support/snyk-for-javascript

**Reply by ghiscoding** [reply]
that explains it lol, thanks for looking into this. Today I got a PR from Snyk about a package to update in a `__fixtures__`, it's probably because I added the project multiple times and I'm not even sure what config it has now

**Reply by haydenmcp** [reply]
Yep...lol. Hmmm...is it a reasonable recommendation? It should give some indication of why it's a good idea to upgrade to a particular version. You know, you may actually still be able to benefit from usage of Snyk if it's posting PRs. Though, it's probably leaving out your packages which kind of sucks. :-/

**Reply by haydenmcp** [reply]
This is really bizarre. From my fork of lerna-lite I see package.json files from the helpers directory included in the project as well as the root-level package.json. But I wonder why it has a problem with the package.json files in each of the packages while it's finding those in the helpers dir.

**Reply by ghiscoding** [reply]
yeah it's a bit strange, it still doesn't show up in my list of projects even though it created a PR this morning, why I don't know. So I can't even configure it, would it be better to maybe add a snyk config in the project root? I wonder if it will update the lock file properly though since it doesn't support npm workspaces, I wouldn't it to cause more problems. I might switch from dependabot to renovate since we can multiple updates in a single PR which dependabot doesn't support yet, for now dependabot is schedule to run every week so on Monday I always get a few PRs. I might not need Snyk since I already have a maintenance plan.

Oh I just found this page under settings where I can deactivate some checks

![image](https://user-images.githubusercontent.com/643976/171206781-dded335d-bd0a-4948-a3ef-b4cd7550888a.png)

**Reply by haydenmcp** [reply]
Yeah, I'm sorry the process was so strange...it sounds like our accounts are behaving very differently. If it's not the right tool for you, it's not the right tool. Did you happen to click on the projects tab? Is that where nothing shows up?

**Reply by haydenmcp** [reply]
Don't feel the need to waste too much time on it, by the way. Hahaha. I was hoping it'd improve things but it looks like it needs support for npm workspaces.

**Reply by haydenmcp** [reply]
Mind if I take the recommended package change and implement it on my end followed by a PR against your repo?

**Reply by ghiscoding** [reply]
It's ok, I cleaned it up and removed all fixtures and helpers, now I only have 2 lines, I might keep just that and still have dependabot or renovate to do weekly updates (this windows is in Settings -> Usages)

![image](https://user-images.githubusercontent.com/643976/171207717-aa3af975-ed40-4195-9346-084329f6041a.png)

Yeah under project, I see nothing now there, I removed my other lib from the list that was showing yesterday and now it's totally empty except for the usage which shows something. Their platform is a little wonky lol

![image](https://user-images.githubusercontent.com/643976/171208101-f97694ac-c22a-40c1-ad97-80bd63b09104.png)

**Reply by haydenmcp** [reply]
Oh dude, do you see the filters on the left of the page? It's likely not showing up because it's being filtered out.

**Reply by ghiscoding** [reply]
you mean this filter, damn that was stupid of me hahahah

![image](https://user-images.githubusercontent.com/643976/171208837-3b576407-bdfa-4531-b399-d069d20c56f5.png)

and now magic happens

![image](https://user-images.githubusercontent.com/643976/171209014-7ef504c2-ff62-4a90-9a75-b75eb076ce57.png)

**Reply by haydenmcp** [reply]
Hahaha, that's all right. It happens to everyone. That is it but it still very different from mine. Even without the fixtures I have roughly 15 package.json files. And the root level has a high ranked vulnerability. Did you happen to merge any PRs related to vulnerabilities? Maybe you already took care of it.

**Reply by haydenmcp** [reply]
![image](https://user-images.githubusercontent.com/8798357/171210384-6e307102-d255-4f76-a0ad-a33c83b30d64.png)

**Reply by haydenmcp** [reply]
![image](https://user-images.githubusercontent.com/8798357/171210525-52b95cc0-8089-430f-826d-7dcf9f54ac66.png)

**Reply by ghiscoding** [reply]
Thanks for all the help, I did merge a few PRs yesterday addressing some vulnerabilities and also did a version release yesterday. Also the `npm audit` was show 0 vulnerabilities though I think there was 1 or 2 before I merged all PRs. Could it be because you're still pointing to an older version that is before yesterday? New version of Lerna-Lite is `1.4.0` since yesterday... oh yeah I see your test was ran 20 hours ago, that's probably why

**Reply by haydenmcp** [reply]
When I click on the vulnerability I see this...
![image](https://user-images.githubusercontent.com/8798357/171210736-dbc0611e-9c2a-4cda-922f-06e8d9044378.png)

**Reply by haydenmcp** [reply]
Aha, let me try to point to the new version.

**Reply by haydenmcp** [reply]
Yep, it's fixed! :-)

**Reply by ghiscoding** [reply]
Great, thanks again for this great training session 😉 I think I'm all set with what I currently have, we should be all good to keep it secure 😄

**Reply by haydenmcp** [reply]
Hahaha, you're welcome. Thanks for continuing development of lerna! I'm very happy it's being modernized and those vulnerabilities fixed. :-)
